spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala

```diff
@@ -27,86 +27,153 @@ import org.apache.hadoop.fs.{Path, PathFilter}
 import org.apache.spark.sql.catalyst.TableIdentifier
-import org.apache.spark.sql.catalyst.catalog.CatalogTablePartition
+import org.apache.spark.sql.catalyst.analysis.UnresolvedAttribute
+import org.apache.spark.sql.catalyst.catalog.{CatalogTable, CatalogTablePartition}
+import org.apache.spark.sql.catalyst.expressions.Expression
 import scala.collection.JavaConverters._
+import scala.util.Try
 
 object SparkTableUtil {
   /**
-   * Returns a DataFrame with a row for each partition in the table.
-   *
-   * The DataFrame has 3 columns, partition key (a=1/b=2), partition location, and format
-   * (avro or parquet).
+   * Returns all partitions in the table.
    *
    * @param spark a Spark session
    * @param table a table name and (optional) database
-   * @return a DataFrame of the table's partitions
+   * @return all table's partitions
    */
-  def partitionDF(spark: SparkSession, table: String): DataFrame = {
-    import spark.implicits._
+  def getPartitions(spark: SparkSession, table: String): Seq[SparkPartition] = {
+    val tableIdentifier = spark.sessionState.sqlParser.parseTableIdentifier(table)
+    getPartitions(spark, tableIdentifier)
+  }
 
-    val partitions: Seq[(Map[String, String], Option[String], Option[String])] =
-      Hive.partitions(spark, table).map { p: CatalogTablePartition =>
-        (p.spec, p.storage.locationUri.map(String.valueOf(_)), p.storage.serde)
-      }
+  /**
+   * Returns all partitions in the table.
+   *
+   * @param spark a Spark session
+   * @param tableIdent a table identifier
+   * @return all table's partitions
+   */
+  def getPartitions(spark: SparkSession, tableIdent: TableIdentifier): Seq[SparkPartition] = {
+    val catalog = spark.sessionState.catalog
+    val catalogTable = catalog.getTableMetadata(tableIdent)
 
-    partitions.toDF("partition", "uri", "format")
+    catalog
+      .listPartitions(tableIdent)
+      .map(catalogPartition => toSparkPartition(catalogPartition, catalogTable))
   }
 
   /**
-    * Returns a DataFrame with a row for each partition that matches the specified 'expression'.
-    *
-    * @param spark a Spark session.
-    * @param table name of the table.
-    * @param expression The expression whose matching partitions are returned.
-    * @return a DataFrame of the table partitions.
-    */
-  def partitionDFByFilter(spark: SparkSession, table: String, expression: String): DataFrame = {
-    import spark.implicits._
+   * Returns partitions that match the specified 'predicate'.
+   *
+   * @param spark a Spark session
+   * @param table a table name and (optional) database
+   * @param predicate a predicate on partition columns
+   * @return matching table's partitions
+   */
+  def getPartitionsByFilter(spark: SparkSession, table: String, predicate: String): Seq[SparkPartition] = {
+    val tableIdentifier = spark.sessionState.sqlParser.parseTableIdentifier(table)
+
+    val unresolvedPredicateExpr = spark.sessionState.sqlParser.parseExpression(predicate)
```

Comment:
I don't like the need to resolve the attributes manually, but it is safer than using the hive client directly (not all versions actually support native filtering, for example).